contracts/amm/pool/Joseph.sol

```diff
@@ -92,62 +100,74 @@ abstract contract Joseph is JosephInternal, IJoseph {
     }
 
     function _redeem(uint256 ipTokenAmount, uint256 timestamp) internal nonReentrant {
+        IIpToken ipToken = _getIpToken();
         require(
-            ipTokenAmount > 0 && ipTokenAmount <= _getIpToken().balanceOf(_msgSender()),
+            ipTokenAmount > 0 && ipTokenAmount <= ipToken.balanceOf(_msgSender()),
             JosephErrors.CANNOT_REDEEM_IP_TOKEN_TOO_LOW
         );
         IMiltonInternal milton = _getMilton();
 
-        uint256 exchangeRate = _calculateExchangeRate(timestamp);
+        IporTypes.MiltonBalancesMemory memory balance = milton.getAccruedBalance();
 
-        require(exchangeRate > 0, MiltonErrors.LIQUIDITY_POOL_IS_EMPTY);
+        uint256 exchangeRate = _calculateExchangeRate(
+            timestamp,
+            milton,
+            ipToken,
+            balance.liquidityPool
+        );
 
-        uint256 wadAssetAmount = IporMath.division(ipTokenAmount * exchangeRate, Constants.D18);
+        require(exchangeRate > 0, MiltonErrors.LIQUIDITY_POOL_IS_EMPTY);
 
-        uint256 wadRedeemFee = IporMath.division(
-            wadAssetAmount * _getRedeemFeeRate(),
-            Constants.D18
+        JosephTypes.RedeemMoney memory redeemMoney = _calculateRedeemMoney(
+            ipTokenAmount,
+            exchangeRate
         );
 
-        uint256 redeemAmount = IporMath.convertWadToAssetDecimals(
-            wadAssetAmount - wadRedeemFee,
+        uint256 wadMiltonErc20Balance = IporMath.convertToWad(
+            IERC20Upgradeable(_getAsset()).balanceOf(address(milton)),
```

Comment:
You read this two times from storage lines 127 and 157 
could it be cheaper to read it once?

Reply:
It was because of stack is too deep, but I used now new structure RedeemMoney so now I can save asset in variable. I will change it.

Reply:
Done.